lib/LifecycleHooks.lua

```diff
@@ -0,0 +1,168 @@
+local TestEnum = require(script.Parent.TestEnum)
+local Stack = require(script.Parent.Stack)
+
+local LifecycleHooks = {}
+LifecycleHooks.__index = LifecycleHooks
+
+function LifecycleHooks.new()
+	local self = {}
+	setmetatable(self, LifecycleHooks)
+	self._stack = Stack.new()
+	return self
+end
+
+--[[
+	Returns an array of `beforeEach` hooks in FIFO order
+]]
+function LifecycleHooks:getBeforeEachHooks()
+	local key = TestEnum.NodeType.BeforeEach
+	local hooks = {}
+
+	for _, level in ipairs(self._stack.data) do
+		for _, hook in ipairs(level[key]) do
+			table.insert(hooks, hook)
+		end
+	end
+
+	return hooks
+end
+
+--[[
+	Returns an array of `afterEach` hooks in FILO order
+]]
+function LifecycleHooks:getAfterEachHooks()
+	local key = TestEnum.NodeType.AfterEach
+	local hooks = {}
+	for _, level in ipairs(self._stack.data) do
+		for _, hook in ipairs(level[key]) do
+			table.insert(hooks, 1, hook)
+		end
+	end
+
+	return hooks
+end
+
+--[[
+	Pushes uncalled beforeAll and afterAll hooks back up the stack
+]]
+function LifecycleHooks:popHooks()
+	local popped = self._stack:pop()
+
+	local function pushHooksUp(type)
+
+		local back = self:_getBackOfStack()
+
+		if not back then
+			return
+		end
+
+		back[type] = popped[type]
+	end
+
+	pushHooksUp(TestEnum.NodeType.BeforeAll)
+	pushHooksUp(TestEnum.NodeType.AfterAll)
+end
+
+function LifecycleHooks:pushHooksFrom(planNode)
+	assert(planNode ~= nil)
+
+	self._stack:push({
+		[TestEnum.NodeType.BeforeAll] = self:_getBeforeAllHooksUncalledAtCurrentLevel(planNode.children),
+		[TestEnum.NodeType.AfterAll] = self:_getAfterAllHooksUncalledAtCurrentLevel(planNode.children),
+		[TestEnum.NodeType.BeforeEach] = self:_getHooksOfType(planNode.children, TestEnum.NodeType.BeforeEach),
+		[TestEnum.NodeType.AfterEach] = self:_getHooksOfType(planNode.children, TestEnum.NodeType.AfterEach),
+	})
+end
+
+function LifecycleHooks:getPendingBeforeHooks()
+	return self:_getAndClearPendingHooks(TestEnum.NodeType.BeforeAll)
+end
+
+function LifecycleHooks:getAfterAllHooks()
+	if self._stack:size() > 0 then
+		return self:_getAndClearPendingHooks(TestEnum.NodeType.AfterAll)
+	else
+		return {}
+	end
+end
+
+--[[
+	Return any hooks that have not yet been returned for this key and clear those hooks
+]]
+function LifecycleHooks:_getAndClearPendingHooks(key)
+	assert(key ~= nil)
+
+	if self._stack:size() > 0 then
+
+		local back = self._stack:getBack()
+
+		local hooks = back[key]
+
+		back[key] = {}
+
+		return hooks
+	else
+		return {}
+	end
```

Comment:
A lot of this code gets to be a lot simpler if we switch from `Stack` to a plain table, since we won't error trying to do things like get the back of the stack.